### ISSUE TITLE
Conversion problem?

### ISSUE THREAD

**Issue by unhandyandy** [JuliaPOMDP/MCTS.jl#39]
I'm 99% sure this is my fault, but I'm just trying out the package, and everything seems to work so far except.

```
action(policy, startpos)

MethodError: Cannot `convert` an object of type Array{MCTS.StateActionNode{TTTMove},2} to an object of type Array{MCTS.StateActionNode{TTTMove},1}
This may have arisen from a call to the constructor Array{MCTS.StateActionNode{TTTMove},1}(...),
since type constructors fall back to convert methods.
```
I.e. the solver and policy are defined, but I get this error when testing the policy.

**Comment by zsunberg** [comment]
Fascinating... can you post the rest of the error (i.e. the line numbers)?

It would also be helpful to see all of your code if you're willing to post it.

**Comment by unhandyandy** [comment]
Here's the full error message.

```
INFO: Loading HttpServer methods...

LoadError: MethodError: Cannot `convert` an object of type Array{MCTS.StateActionNode{TTTMove},2} to an object of type Array{MCTS.StateActionNode{TTTMove},1}
This may have arisen from a call to the constructor Array{MCTS.StateActionNode{TTTMove},1}(...),
since type constructors fall back to convert methods.
while loading /home/dabrowsa/lang/julia/mine/mcts-ttt.jl, in expression starting on line 155

Stacktrace:
 [1] MCTS.StateNode(::MCTS.MCTSPlanner{TTTGame,TTTState,TTTMove,MCTS.SolvedRolloutEstimator{POMDPToolbox.RandomPolicy{MersenneTwister,TTTGame,POMDPToolbox.VoidUpdater},MersenneTwister},MersenneTwister}, ::TTTState) at /home/dabrowsa/.julia/v0.6/MCTS/src/vanilla.jl:22
 [2] insert_node!(::MCTS.MCTSPlanner{TTTGame,TTTState,TTTMove,MCTS.SolvedRolloutEstimator{POMDPToolbox.RandomPolicy{MersenneTwister,TTTGame,POMDPToolbox.VoidUpdater},MersenneTwister},MersenneTwister}, ::TTTState) at /home/dabrowsa/.julia/v0.6/MCTS/src/vanilla.jl:206
 [3] simulate(::MCTS.MCTSPlanner{TTTGame,TTTState,TTTMove,MCTS.SolvedRolloutEstimator{POMDPToolbox.RandomPolicy{MersenneTwister,TTTGame,POMDPToolbox.VoidUpdater},MersenneTwister},MersenneTwister}, ::TTTState, ::Int64) at /home/dabrowsa/.julia/v0.6/MCTS/src/vanilla.jl:164
 [4] build_tree(::MCTS.MCTSPlanner{TTTGame,TTTState,TTTMove,MCTS.SolvedRolloutEstimator{POMDPToolbox.RandomPolicy{MersenneTwister,TTTGame,POMDPToolbox.VoidUpdater},MersenneTwister},MersenneTwister}, ::TTTState) at /home/dabrowsa/.julia/v0.6/MCTS/src/vanilla.jl:141
 [5] action(::MCTS.MCTSPlanner{TTTGame,TTTState,TTTMove,MCTS.SolvedRolloutEstimator{POMDPToolbox.RandomPolicy{MersenneTwister,TTTGame,POMDPToolbox.VoidUpdater},MersenneTwister},MersenneTwister}, ::TTTState) at /home/dabrowsa/.julia/v0.6/MCTS/src/vanilla.jl:129
 [6] include_from_node1(::String) at ./loading.jl:576
 [7] include(::String) at ./sysimg.jl:14
```

The code is just a naive attempt to simulate a tic-tac-toe game, based on the GridWorld example.  No doubt there are some errors here.

```
using POMDPs,MCTS,D3Trees

struct TTTState 
    board::Array{Int,2}  # board
    player::Int # on move, 1 or 2
    done::Bool
end

struct TTTMove
    x::Int
    y::Int
end

startpos = TTTState([0 0 0; 0 0 0; 0 0 0],1,false)
testpos1 = TTTState([1 1 1; 2 2 0; 0 0 0],2,true)
testpos2 = TTTState([1 1 0; 2 2 2; 1 0 0],1,true)

POMDPs.isequal(s1::TTTState, s2::TTTState) = s1.board == s2.board && s1.player == s2.player

POMDPs.hash(state::TTTState,n::Int64) = POMDPs.state_index(mdp, state)
POMDPs.hash(state::TTTState) = POMDPs.hash(state,0
                                           )
struct TTTGame <: MDP{TTTState, TTTMove} 
end

POMDPs.actions(::TTTGame) = [TTTMove(i,j) for i=1:3, j=1:3 ];

function inbounds(mdp::TTTGame,x::Int,y::Int)::Boolean
    return 1 <= x <= 3 && 1 <= y <= 3 && mdp.board[x,y]==0
end

inbounds(mdp::TTTGame, move::TTTMove) = inbounds(mdp, move.x, move.y);

function POMDPs.transition(mdp::TTTGame, state::TTTState, move::TTTMove)::TTTGame
    if !inbounds(mdp, move)
        # If would transition out of bounds, stay in
        # same cell with probability 1
        return state
    else
        player = state.player
        newboard = state.board
        newboard[move.x,move.y] = player
        return TTTGame(newboard,3 - player)
    end
end;

lines = [ [ [1, 1], [2, 2], [3, 3] ],
          [ [3, 1], [2, 2], [1, 3] ],
          [ [3, 1], [3, 2], [3, 3] ],
          [ [2, 1], [2, 2], [2, 3] ],
          [ [1, 1], [1, 2], [1, 3] ],
          [ [3, 3], [2, 3], [1, 3] ],
          [ [3, 2], [2, 2], [1, 2] ],
          [ [3, 1], [2, 1], [1, 1] ], ];

function check_constant( l::Array{Int,1})::Bool
    h, t = l[1], l[2:3]
    test = map( x -> x==h, t)
    return reduce(&,test)
end

function check_one_line(state::TTTState,l::Array{Array{Int,1},1})::Int
    b = state.board
    pces = map(c -> b[c...], l)
    check_constant(pces) ? (return pces[1]) : (return 0)
end

function check_lines(state::TTTState)::Int
    b = state.board
    p, q = state.player, 3 - state.player
    tests = map( l -> check_one_line(state,l), lines)
    return (q in tests) ? q : (p in tests) ? p : 0
end  


function POMDPs.reward(mdp::TTTGame, state::TTTState, move::TTTMove, statep::TTTState)
    p = statep.player
    winner = check_lines(statep)
    ## Signs??
    return (winner==p) ? -1 : (winner == 3 - p) ? 1 : 0
end;



POMDPs.n_states(mdp::TTTGame) = 3^9
POMDPs.n_actions(mdp::TTTGame) = 9

POMDPs.discount(mdp::TTTGame) = 0;

function POMDPs.state_index(mdp::TTTGame, state::TTTState)
    sd = Int(state.done + 1)
    pces = [ 1 + state.board[ind2sub((3,3),i)...] for i=1:9 ]
    return sub2ind((3,3,3,3,3,3,3,3,3), pces...)
end

function POMDPs.action_index(mdp::TTTGame, move::TTTMove)
    return sub2ind((3,3), move.x, move.y)
end;

POMDPs.isterminal(mdp::TTTGame, s::TTTState) = s.done

function MCTS.node_tag(s::TTTState)
    return "$s.board"
end

MCTS.node_tag(m::TTTMove) = "[$m]"

n_iter = 100000
depth = 9
ec = 10.0

solver = MCTSSolver(n_iterations=n_iter,
    depth=depth,
    exploration_constant=ec,
    enable_tree_vis=true
)
mdp = TTTGame(3)

policy = solve(solver, mdp)

a = action(policy, startpos)

```

**Comment by zsunberg** [comment]
Hi @unhandyandy ,

The problem stems from the fact that your `actions()` function returns a matrix rather than just a 1D array.  I think that's an annoying thing about how Julia generators work.

Leave the issue open, I think I'll try to catch this in the MCTS code if possible.

- Zach

**Comment by unhandyandy** [comment]
So I was right - it was my fault.  Thanks for looking unto it. 

Are you keeping it open because you think the rank 2 array should be coerced to a list?  My main complaint is just that the error message (even with the stack) was so opaque I didn't know in what direction to look.

**Comment by zsunberg** [comment]
> So I was right - it was my fault. Thanks for looking unto it.

Eh, arguably not your fault - the only stipulation that we have is that the
thing returned by `iterator(actions(...))` should be iterable, so you met
that requirement. I just didn't think about this possibility when writing
the MCTS code.

Yeah, I think I'm just going to coerce it (
https://github.com/JuliaPOMDP/MCTS.jl/pull/40), there shouldn't be
significant additional overhead with `vec()` because it uses the same
memory. Can you think of a better way? Eventually, it would be best to have
some kind of a POMDPs.jl linter that looks at problems and helps people
find errors, but I have to finish my thesis

**Comment by unhandyandy** [comment]
So it should really not just be iterable, but of type something like

Array{Action,1}

That doesn't seem unreasonable.

**Comment by zsunberg** [comment]
No, I think the best requirement is just to have it be iterable and *not*
require it to be an Array :)

The solver should be able to handle anything that's iterable - I believe it
is fixed now after merging #40. If the solver wants to collect all the
actions in memory (as I am doing in MCTS), that's fine, but we shouldn't
require the allocation of an Array. There are many cases where it is
beneficial to skip the allocation and just iterate through it.

Also, a tip for your future Julia coding: usually you want to write things
that can handle AbstractArray{Action, 1} instead of Array{Action, 1}. That
way people can use really powerful things like StaticArrays (
https://github.com/JuliaArrays/StaticArrays.jl)
